pkg/kp/consulutil/watch.go

```diff
@@ -0,0 +1,165 @@
+package consulutil
+
+import (
+	"time"
+
+	"github.com/square/p2/Godeps/_workspace/src/github.com/hashicorp/consul/api"
+)
+
+// WatchPrefix watches a Consul prefix for changes to any keys that have the prefix. When
+// anything changes, all Key/Value pairs having that prefix will be written to a provided
+// channel.
+//
+// Errors will sent on the given output channel but do not otherwise affect execution. The
+// provided output stream will become owned by this function call, and this call will
+// close it when the function ends. This function will run until explicitly canceled by
+// closing the "done" channel. Data is written to the output channel synchronously, so
+// readers must consume the data to this method will block.
+func WatchPrefix(
+	prefix string,
+	clientKV ConsulLister,
+	outPairs chan<- api.KVPairs,
+	done <-chan struct{},
+	outErrors chan<- error,
+) {
+	defer close(outPairs)
+	var currentIndex uint64
+	timer := time.NewTimer(time.Duration(0))
+
+	for {
+		select {
+		case <-done:
+			return
+		case <-timer.C:
+		}
+		timer.Reset(250 * time.Millisecond) // upper bound on request rate
+		pairs, queryMeta, err := SafeList(clientKV, done, prefix, &api.QueryOptions{
+			WaitIndex: currentIndex,
+		})
+		switch err {
+		case CanceledError:
+			return
+		case nil:
+			currentIndex = queryMeta.LastIndex
+			select {
+			case <-done:
+			case outPairs <- pairs:
+			}
+		default:
+			select {
+			case <-done:
+			case outErrors <- err:
```

Comment:
**Short answer**: Let's save it for another diff.

**Long answer:**

If only Golang had lightweight (syntax-wise) anonymous tuples.

I too have been thinking a lot about how we handle errors, and I actually like both approaches in different circumstances.

(1) works well when the errors are meaningful and fatal and you really need a higher layer to handle them. Imagine something that produces a sequence of values and then hits an error that prevents it from going any further. This would be the most direct way to build something that treats channels like an asynchronous iterator/stream with exception-like behavior (e.g., a Golang version of [Rx](http://reactivex.io/) in Golang or perhaps an Exception Monad in Haskell).

(2) works well when the errors really don't matter.  The method is designed to just keep going in the presence of most errors.  But for debugging and metrics, you still want to know, somehow, that the errors were occurring, so you log them.

Getting back to your question, (2) seems more suitable for this kind of watcher, and my early versions threaded a `logging.Logger` through many of the methods.  I don't like the error channel.  I think it's an anti-pattern that we've repeated everywhere.  But once I started trying to integrate this into our existing code, I took out the loggers and went back to the error channels.  That pattern is too deep, and this change will be big enough already.  Changing all `chan<- error` arguments into `logging.Logger` needs its own PR.